lib/features/auto-place/AutoPlaceUtil.js

```diff
@@ -109,133 +113,124 @@ export function getConnectedAtPosition(source, position, element) {
+        return asTRBL(b).top - asTRBL(a).bottom;
+      } else if (reference === 'center') {
+        return getMid(b).y - asTRBL(a).bottom;
+      } else {
+        return asTRBL(b).bottom - asTRBL(a).bottom;
+      }
+    } else {
+      if (reference === 'start') {
+        return asTRBL(a).left - asTRBL(b).right;
+      } else if (reference === 'center') {
+        return asTRBL(a).left - getMid(b).x;
+      } else {
+        return asTRBL(a).left - asTRBL(b).left;
+      }
+    }
+  }
 
-    return function(shape) {
+  var sourcesDistances = source.incoming
+    .filter(filter)
+    .map(function(connection) {
       return {
-        shape: shape,
-        weight: weight,
-        distanceTo: function(shape) {
-          var shapeTrbl = asTRBL(shape);
-
-          if (axis === 'x') {
-            return shapeTrbl.left - sourceTrbl.right;
-          } else {
-            return shapeTrbl.top - sourceTrbl.bottom;
-          }
-        }
+        id: connection.source.id,
+        distance: getDistance(source, connection.source),
+        weight: getWeight(connection)
       };
-    };
-  }
+    });
 
-  function toSourceNode(weight) {
-    return function(shape) {
+  var targetsDistances = source.outgoing
+    .filter(filter)
+    .map(function(connection) {
       return {
-        shape: shape,
-        weight: weight,
-        distanceTo: function(shape) {
-          var shapeTrbl = asTRBL(shape);
-
-          if (axis === 'x') {
-            return sourceTrbl.left - shapeTrbl.right;
-          } else {
-            return sourceTrbl.top - shapeTrbl.bottom;
-          }
-        }
+        id: connection.target.id,
+        distance: getDistance(source, connection.target),
+        weight: getWeight(connection)
       };
-    };
-  }
-
-  // we create a list of nodes to take into consideration
-  // for calculating the optimal flow node distance
-  //
-  //   * weight existing target nodes higher than source nodes unless otherwise indicated
-  //   * only take into account individual nodes once
-  //
-  var nodes = reduce([].concat(
-    getTargets(source, filter).map(connectionTargetIsSource ? toSourceNode(5) : toTargetNode(5)),
-    getSources(source, filter).map(connectionTargetIsSource ? toTargetNode(1) : toSourceNode(1))
-  ), function(nodes, node) {
-
-    // filter out shapes connected twice via source or target
-    nodes[node.shape.id + '__weight_' + node.weight] = node;
-
-    return nodes;
-  }, {});
+    });
 
-  // compute distances between source and incoming nodes;
-  // group at the same time by distance and expose the
-  // favourite distance as { fav: { count, value } }.
-  var distancesGrouped = reduce(nodes, function(result, node) {
-    var shape = node.shape,
-        weight = node.weight,
-        distanceTo = node.distanceTo;
+  var distances = sourcesDistances.concat(targetsDistances).reduce(function(accumulator, currentValue) {
+    accumulator[ currentValue.id + currentValue.weight ] = currentValue;
```

Comment:
Although numerical ids cause a crash in diagram-js, they are not openly disallowed. Thus, I'd suggest to explicitly make sure we are handling strings here, like it used to be:
```suggestion
    accumulator[ currentValue.id + '__weight_' + currentValue.weight ] = currentValue;
```

Reply:
Good catch. Thanks.